subprojects/messaging/src/main/java/org/gradle/internal/serialize/DefaultSerializerRegistry.java

```diff
@@ -19,24 +19,32 @@
 import com.google.common.base.Objects;
 import org.gradle.internal.Cast;
 
+import javax.annotation.concurrent.ThreadSafe;
 import java.util.Arrays;
 import java.util.Comparator;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
 import java.util.Map;
 import java.util.Set;
-import java.util.TreeMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.ConcurrentSkipListSet;
 
+/**
+ * Default implementation of {@link SerializerRegistry}.
+ *
+ * This class must be thread-safe because multiple tasks may be registering serializable classes concurrently, while other tasks are calling {@link #build(Class)}.
+ */
+@ThreadSafe
 public class DefaultSerializerRegistry implements SerializerRegistry {
-    private final Map<Class<?>, Serializer<?>> serializerMap = new TreeMap<Class<?>, Serializer<?>>(new Comparator<Class<?>>() {
+    private static final Comparator<Class<?>> CLASS_COMPARATOR = new Comparator<Class<?>>() {
         @Override
         public int compare(Class<?> o1, Class<?> o2) {
             return o1.getName().compareTo(o2.getName());
         }
-    });
-    private final Set<Class<?>> javaSerialization = new HashSet<Class<?>>();
+    };
+    private final Map<Class<?>, Serializer<?>> serializerMap = new ConcurrentSkipListMap<Class<?>, Serializer<?>>(CLASS_COMPARATOR);
+    private final Set<Class<?>> javaSerialization = new ConcurrentSkipListSet<Class<?>>(CLASS_COMPARATOR);
```

Comment:
Do you remember why you switched to a set based on the comparator here? I feel a little bit uneasy about this since it changes the behaviour: `equals` and `compare` of the comparator are not consistent: `compare() == 0` even though `equals() == false` when the same class is loaded from a different classloader. That is why I think that this `Set` should not use `ConcurrentSkipListSet` but something without an ordering - something like `CopyOnWriteArraySet` or a set backed by `ConcurrentHashSet` would be good I think.

WDYT?